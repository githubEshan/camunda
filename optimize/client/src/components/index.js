/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */

export {Analysis} from './Analysis';
export {Header} from './Header';
export {PrivateRoute} from './PrivateRoute';
export {Dashboard} from './Dashboards';
export {Report} from './Reports';
export {Sharing} from './Sharing';
export {Home, Collection} from './Home';
export {default as Logout} from './Logout';
export {Processes, ProcessReport} from './Processes';
