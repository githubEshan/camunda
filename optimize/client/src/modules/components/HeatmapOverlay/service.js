/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */

import HeatmapJS from 'heatmap.js';
import ReactDOM from 'react-dom';

const SEQUENCEFLOW_RADIUS = 30;
const SEQUENCEFLOW_STEPWIDTH = 10;
const SEQUENCEFLOW_VALUE_MODIFIER = 0.2;
const ACTIVITY_DENSITY = 20;
const ACTIVITY_RADIUS = 50;
const ACTIVITY_VALUE_MODIFIER = 0.125;
const VALUE_SHIFT = 0.17;
const COOLNESS = 2.5;
const EDGE_BUFFER = 75;
const RESOLUTION = 4;

export function getHeatmap(viewer, data, noSequenceHighlight) {
  const heat = generateHeatmap(viewer, data, noSequenceHighlight);
  const node = document.createElementNS('http://www.w3.org/2000/svg', 'image');

  Object.keys(heat.dimensions).forEach((prop) => {
    node.setAttributeNS(null, prop, heat.dimensions[prop]);
  });

  node.setAttributeNS('http://www.w3.org/1999/xlink', 'xlink:href', heat.img);
  node.setAttributeNS(null, 'style', 'opacity: 0.8; pointer-events: none;');

  return node;
}

function generateHeatmap(viewer, data, noSequenceHighlight) {
  const dimensions = getDimensions(viewer);
  let heatmapData = generateData(data, viewer, dimensions, noSequenceHighlight);

  const map = createMap(dimensions);
  const heatmapDataValueMax = Math.max.apply(
    this,
    heatmapData.map((el) => el.value)
  );

  heatmapData = heatmapData.map(({x, y, value, radius}) => {
    const shiftValue = noSequenceHighlight ? 0 : VALUE_SHIFT;
    return {
      x: Math.round(x),
      y: Math.round(y),
      radius,
      value: (shiftValue + (value / heatmapDataValueMax) * (1 - shiftValue)) / COOLNESS,
    };
  });

  map.setData({
    min: 0,
    max: 1,
    data: heatmapData,
  });

  return {
    img: map.getDataURL(),
    dimensions,
  };
}

function getDimensions(viewer) {
  const dimensions = viewer.get('canvas').getActiveLayer().getBBox();

  return {
    width: dimensions.width + 2 * EDGE_BUFFER,
    height: dimensions.height + 2 * EDGE_BUFFER,
    x: dimensions.x - EDGE_BUFFER,
    y: dimensions.y - EDGE_BUFFER,
  };
}

function createMap(dimensions) {
  const container = document.createElement('div');

  container.style.width = dimensions.width / RESOLUTION + 'px';
  container.style.height = dimensions.height / RESOLUTION + 'px';
  container.style.position = 'absolute';

  document.body.appendChild(container);

  const map = HeatmapJS.create({container});

  document.body.removeChild(container);

  container.firstChild.setAttribute('width', dimensions.width / RESOLUTION);
  container.firstChild.setAttribute('height', dimensions.height / RESOLUTION);

  return map;
}

function isBpmnType(element, types) {
  if (typeof types === 'string') {
    types = [types];
  }
  return (
    element.type !== 'label' &&
    types.filter((type) => element.businessObject.$instanceOf('bpmn:' + type)).length > 0
  );
}

function isExcluded(element) {
  return isBpmnType(element, 'SubProcess') && !element.collapsed;
}

function generateData(values, viewer, {x: xOffset, y: yOffset}, noSequenceHighlight) {
  const data = [];
  const elementRegistry = viewer.get('elementRegistry');

  for (const key in values) {
    const element = elementRegistry.get(key);

    if (!element || typeof values[key] !== 'number') {
      // for example for multi instance bodies
      continue;
    }

    if (!isExcluded(element)) {
      // add multiple points evenly distributed in the area of the node. Number of points depends on area of node
      for (let i = 0; i < element.width + ACTIVITY_DENSITY / 2; i += ACTIVITY_DENSITY) {
        for (let j = 0; j < element.height + ACTIVITY_DENSITY / 2; j += ACTIVITY_DENSITY) {
          const value = values[key] === 0 ? Number.EPSILON : values[key];

          data.push({
            x: (element.x + i - xOffset) / RESOLUTION,
            y: (element.y + j - yOffset) / RESOLUTION,
            value: value * ACTIVITY_VALUE_MODIFIER,
            radius: ACTIVITY_RADIUS / RESOLUTION,
          });
        }
      }
    }

    if (!noSequenceHighlight) {
      for (let i = 0; i < element.incoming.length; i++) {
        drawSequenceFlow(
          data,
          element.incoming[i].waypoints,
          Math.min(values[key], values[element.incoming[i].source.id]),
          {xOffset, yOffset}
        );
      }
    }
  }

  return data;
}

function drawSequenceFlow(data, waypoints, value, {xOffset, yOffset}) {
  if (!value) {
    return;
  }

  for (let i = 1; i < waypoints.length; i++) {
    const start = waypoints[i - 1];
    const end = waypoints[i];

    const movementVector = {
      x: end.x - start.x,
      y: end.y - start.y,
    };
    const normalizedMovementVector = {
      x:
        (movementVector.x / (Math.abs(movementVector.x) + Math.abs(movementVector.y))) *
        SEQUENCEFLOW_STEPWIDTH,
      y:
        (movementVector.y / (Math.abs(movementVector.x) + Math.abs(movementVector.y))) *
        SEQUENCEFLOW_STEPWIDTH,
    };

    const numberSteps =
      Math.sqrt(movementVector.x * movementVector.x + movementVector.y * movementVector.y) /
      SEQUENCEFLOW_STEPWIDTH;

    for (let j = 0; j < numberSteps; j++) {
      data.push({
        x: (start.x + normalizedMovementVector.x * j - xOffset) / RESOLUTION,
        y: (start.y + normalizedMovementVector.y * j - yOffset) / RESOLUTION,
        value: value * SEQUENCEFLOW_VALUE_MODIFIER,
        radius: SEQUENCEFLOW_RADIUS / RESOLUTION,
      });
    }
  }
}

export async function addDiagramTooltip({
  viewer,
  element,
  tooltipContent,
  theme,
  onMouseEnter,
  onMouseLeave,
}) {
  const elementGraphics = viewer.get('elementRegistry').getGraphics(element);
  const overlaysContainer = document.body.querySelector('.djs-overlay-container');
  if (!elementGraphics || !tooltipContent || !overlaysContainer) {
    return;
  }

  // create overlay node from html
  const overlayHtml = document.createElement('div');
  overlayHtml.classList.add('Tooltip', theme === 'light' ? 'light' : 'dark');

  if (onMouseEnter) {
    overlayHtml.addEventListener('mouseenter', onMouseEnter);
  }

  if (onMouseLeave) {
    overlayHtml.addEventListener('mouseleave', onMouseLeave);
  }

  // render tooltip react markup into the html tooltip container
  await insertReactContent(<div className="tooltipBridge">{tooltipContent}</div>, overlayHtml);
  overlaysContainer.appendChild(overlayHtml);
  const overlayHeight = overlayHtml.clientHeight;
  overlaysContainer.removeChild(overlayHtml);

  const nodeWidth = elementGraphics.querySelector('.djs-visual').getBBox().width;
  const position = {
    left: nodeWidth / 2,
  };

  if (
    viewer.get('elementRegistry').get(element).y - viewer.get('canvas').viewbox().y <
    overlayHeight
  ) {
    position.bottom = 0;
    const classList = overlayHtml.classList;
    classList.remove('top');
    classList.add('bottom');
  } else {
    position.top = 0;
  }

  // add overlay to viewer
  return viewer.get('overlays').add(element, 'TOOLTIP', {
    position,
    show: {
      minZoom: -Infinity,
      maxZoom: +Infinity,
    },
    html: overlayHtml,
  });
}

function insertReactContent(tooltipContent, overlayHtml) {
  return new Promise((resolve) => {
    ReactDOM.render(tooltipContent, overlayHtml, resolve);
  });
}
