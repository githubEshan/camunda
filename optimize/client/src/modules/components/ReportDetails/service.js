/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */

import {post} from 'request';
import { getFullURL } from '../../api';

export async function loadTenants(definitions) {
  const params = {definitions};
  const response = await post(getFullURL(`api/definition/process/_resolveTenantsForVersions`), params);

  return await response.json();
}
