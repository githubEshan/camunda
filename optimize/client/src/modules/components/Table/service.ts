/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */

import {CSSProperties} from 'react';
import {Column, UseSortByColumnProps, UseSortByOptions} from 'react-table';

import {t} from 'translation';

import {Head} from './Table';

type Entry = Head;
export function flatten(ctx: string = '', suffix: (entry: Entry) => string | undefined = () => '') {
  return (flat: string[], entry: Entry): string[] => {
    if (typeof entry === 'object' && entry.columns) {
      // nested column, flatten recursivly with augmented context
      return flat.concat(
        entry.columns.reduce<string[]>(flatten(ctx + (entry.id || entry.label), suffix), [])
      );
    } else {
      // normal column, return current context with optional suffix
      return flat.concat(ctx + suffix(entry));
    }
  };
}

// We have to do this, because when header is sortable, style props are only passed to the button that is rendered inside the header
// see implementation here: https://github.com/carbon-design-system/carbon/blob/main/packages/react/src/components/DataTable/TableHeader.tsx#L179
export function rewriteHeaderStyles(styles?: CSSProperties) {
  return function (th: HTMLTableCellElement | null) {
    if (!th) {
      return;
    }

    if (styles) {
      Object.entries(styles).forEach(([key, value]) => {
        // @ts-expect-error this simply rewrites the style
        th.style[key] = value;
      });
    }
  };
}

export function formatSorting<T extends object>(
  sorting: {by: string; order: string} | undefined,
  resultType: string | undefined,
  columns: (Column & Partial<UseSortByOptions<T> & UseSortByColumnProps<T>>)[],
  allowLocalSorting: boolean
): {id?: string; desc?: boolean; order?: string}[] {
  if (allowLocalSorting) {
    const firstSortableColumn = columns.find((column) => !column.disableSortBy);
    if (firstSortableColumn) {
      return [{id: firstSortableColumn.id, desc: false}];
    }
    return [];
  }

  if (!sorting) {
    return [];
  }
  const {by, order} = sorting;
  let id = by;
  if (resultType === 'map') {
    if (columns[0]?.id && (by === 'label' || by === 'key')) {
      id = columns[0]?.id;
    } else if (columns[1]?.id && by === 'value') {
      id = columns[1]?.id;
    }
  }
  return [{id, desc: order === 'desc'}];
}

export function convertHeaderNameToAccessor(name: string) {
  const joined = name
    .split(' ')
    .join('')
    .replace(t('report.variables.default').toString(), t('report.groupBy.variable') + ':');

  return joined.charAt(0).toLowerCase() + joined.slice(1);
}
