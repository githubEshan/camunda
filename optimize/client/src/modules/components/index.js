/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */

export {Table} from './Table';
export {Modal} from './Modal';
export {Deleter} from './Deleter';
export {Select} from './Select';
export {BPMNDiagram, ClickBehavior, PartHighlight} from './BPMNDiagram';
export {CopyToClipboard} from './CopyToClipboard';
export {ReportRenderer, NoDataNotice} from './ReportRenderer';
export {ErrorBoundary} from './ErrorBoundary';
export {ActionItem} from './ActionItem';
export {ShareEntity} from './ShareEntity';
export {DashboardRenderer, DiagramScrollLock} from './DashboardRenderer';
export {DefinitionSelection, VersionPopover, TenantPopover} from './DefinitionSelection';
export {TargetValueBadge} from './TargetValueBadge';
export {ErrorPage} from './ErrorPage';
export {DatePicker} from './DatePicker';
export {ColorPicker} from './ColorPicker';
export {EntityName} from './EntityName';
export {EntityNameForm} from './EntityNameForm';
export {HeatmapOverlay} from './HeatmapOverlay';
export {EntityList} from './EntityList';
export {UserTypeahead} from './UserTypeahead';
export {DateRangeInput} from './DateRangeInput';
export {LastModifiedInfo} from './LastModifiedInfo';
export {ReportDetails} from './ReportDetails';
export {InstanceCount} from './InstanceCount';
export {DocsLink} from './DocsLink';
export {Tabs} from './Tabs';
export {Checklist} from './Checklist';
export {SelectionPreview} from './SelectionPreview';
export {DownloadButton} from './DownloadButton';
export {BulkDeleter} from './BulkDeleter';
export {AlertModal} from './AlertModal';
export {AlertsDropdown} from './AlertsDropdown';
export {DurationChart} from './DurationChart';
export {ReportTemplateModal, DashboardTemplateModal} from './TemplateModal';
export {KpiCreationModal} from './KpiCreationModal';
export {EmptyState} from './EmptyState';
export {PageTitle} from './PageTitle';
export {TextEditor} from './TextEditor';
export {EntityDescription} from './EntityDescription';
export {Popover} from './Popover';
export {InstanceViewTable} from './InstanceViewTable';
export {TenantInfo} from './TenantInfo';
export {TargetSelection} from './TargetSelection';
export {Loading} from './Loading';
export {Breadcrumbs} from './Breadcrumbs';
