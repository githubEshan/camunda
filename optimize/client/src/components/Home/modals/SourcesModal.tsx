/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */

import {useState, useEffect, ChangeEvent} from 'react';
import classnames from 'classnames';
import {
  Button,
  ComboBox,
  TableSelectAll,
  TableSelectRow,
  TableToolbar,
  TableToolbarContent,
  TableToolbarSearch,
} from '@carbon/react';

import {Modal, Table, TenantPopover} from 'components';
import {formatters} from 'services';
import {t} from 'translation';
import {showError} from 'notifications';
import {areTenantsAvailable, getOptimizeProfile} from 'config';
import {useErrorHandling} from 'hooks';
import {Source} from 'types';

import {
  DefinitionWithTenants,
  TenantWithDefinitions,
  getDefinitionsWithTenants,
  getTenantsWithDefinitions,
} from './service';

import './SourcesModal.scss';

const {formatTenantName} = formatters;

interface SourcesModalProps {
  onClose: () => void;
  onConfirm: (definitions?: Source[]) => void;
  confirmText: string;
  preSelectAll?: boolean;
}

export default function SourcesModal({
  onClose,
  onConfirm,
  confirmText,
  preSelectAll,
}: SourcesModalProps): JSX.Element {
  const [definitions, setDefinitions] = useState<DefinitionWithTenants[]>();
  const [tenants, setTenants] = useState<TenantWithDefinitions[]>([]);
  const [selected, setSelected] = useState<Source[]>([]);
  const [selectedTenant, setSelectedTenant] = useState<string | null | undefined>();
  const [query, setQuery] = useState('');
  const [optimizeProfile, setOptimizeProfile] = useState('');
  const {mightFail} = useErrorHandling();

  useEffect(() => {
    mightFail(
      getDefinitionsWithTenants(),
      (definitions) => {
        if (preSelectAll) {
          setSelected(definitions.map(format));
        }
        setDefinitions(definitions);
      },
      showError
    );
  }, [mightFail, preSelectAll]);

  useEffect(() => {
    (async () => {
      const tenantAvailable = await areTenantsAvailable();
      if (tenantAvailable) {
        mightFail(getTenantsWithDefinitions(), setTenants, showError);
        setOptimizeProfile(await getOptimizeProfile());
      }
    })();
  }, [mightFail]);

  const removeExtraTenants = (def: Source): Source => {
    if (typeof selectedTenant === 'undefined') {
      return def;
    }

    return {...def, tenants: [selectedTenant]};
  };

  function createCollection() {
    onConfirm(selected);
  }

  const filteredDefinitions =
    definitions?.filter(
      (def) =>
        definitionHasSelectedTenant(def, selectedTenant) &&
        (def.name || def.key).toLowerCase().includes(query.toLowerCase())
    ) || [];

  function deselectAll() {
    setSelected(
      selected.concat(
        filteredDefinitions
          .filter(({key}) => !selected.some(({definitionKey}) => key === definitionKey))
          .map(format)
          .map(removeExtraTenants)
      )
    );
  }

  function selectAll() {
    setSelected(
      selected.filter(
        ({definitionKey}) => !filteredDefinitions.some(({key}) => definitionKey === key)
      )
    );
  }

  const isInSelected = ({key}: {key?: string}) =>
    selected.some(({definitionKey}) => key === definitionKey);
  const allChecked = filteredDefinitions.every(isInSelected);
  const allIndeterminate = !allChecked && filteredDefinitions.some(isInSelected);

  const tableHead = [
    {
      label: (
        <TableSelectAll
          id="checked"
          key="checked"
          name="checked"
          aria-label="checked"
          className={classnames({hidden: !filteredDefinitions.length})}
          indeterminate={allIndeterminate}
          checked={allChecked}
          onSelect={({target}) => {
            const {checked} = target as HTMLInputElement;
            if (checked) {
              deselectAll();
            } else {
              selectAll();
            }
          }}
        />
      ),
      id: 'checked',
      sortable: false,
      width: 30,
    },
    {label: t('common.name'), id: 'name', sortable: true},
    {label: t('common.entity.type'), id: 'type', sortable: false, width: 80},
  ];

  if (tenants.length !== 0) {
    tableHead.push({
      label: t('common.tenant.label-plural'),
      id: 'tenants',
      sortable: false,
      width: 80,
    });
  }

  const tenantsSelectorItems = [
    {value: undefined, label: t('common.collection.modal.allTenants')},
    ...tenants.map((tenant) => ({
      value: tenant.id,
      label: formatTenantName(tenant),
    })),
  ];

  return (
    <Modal open onClose={onClose} size="lg" className="SourcesModal" isOverflowVisible>
      <Modal.Header title={t('home.sources.add')} />
      <Modal.Content>
        <Table
          toolbar={
            <TableToolbar>
              <TableToolbarContent>
                {tenants.length !== 0 && (
                  <ComboBox
                    id="tenantsSelector"
                    className="tenantsSelector"
                    placeholder={t('common.select').toString()}
                    items={tenantsSelectorItems}
                    itemToString={(tenant) =>
                      (tenant as (typeof tenantsSelectorItems)[number])?.label.toString()
                    }
                    titleText={t('common.tenant.label-plural')}
                    initialSelectedItem={tenantsSelectorItems[0]}
                    onChange={({selectedItem}) => {
                      setSelected([]);
                      setSelectedTenant(selectedItem?.value);
                    }}
                  />
                )}
                <TableToolbarSearch
                  placeholder={t('home.search.name').toString()}
                  onChange={(evt) => {
                    setQuery((evt as ChangeEvent<HTMLInputElement>).target.value);
                  }}
                  onClear={() => {
                    setQuery('');
                  }}
                  expanded
                  data-modal-primary-focus
                />
              </TableToolbarContent>
            </TableToolbar>
          }
          head={tableHead}
          body={filteredDefinitions.map((def) => {
            const selectedDefinition = selected.find(
              ({definitionKey}) => def.key === definitionKey
            );

            const key = def.name || def.key;
            const body = [
              <TableSelectRow
                checked={!!selectedDefinition}
                id={def.key}
                name={key}
                aria-label={key}
                onSelect={({target}) => {
                  const {checked} = target as HTMLInputElement;
                  if (checked) {
                    setSelected([...selected, removeExtraTenants(format(def))]);
                  } else {
                    setSelected((selected) =>
                      selected.filter(({definitionKey}) => def.key !== definitionKey)
                    );
                  }
                }}
              />,
              key,
              def.type || '',
            ];

            if (tenants.length !== 0) {
              if (optimizeProfile === 'ccsm' && tenants.length === 1) {
                const {id, name} = def?.tenants?.[0] || {};
                body.push(<>{formatTenantName({id, name})}</>);
              } else {
                body.push(
                  // clicking inside the popover
                  <TenantPopover
                    tenants={def.tenants}
                    selected={selectedDefinition?.tenants || ['']}
                    disabled={!selectedDefinition}
                    onChange={(newTenants) => {
                      setSelected(
                        selected.map((selectedDefinition) => {
                          if (def.key === selectedDefinition.definitionKey) {
                            return {
                              ...selectedDefinition,
                              tenants:
                                newTenants.length === 0 ? [def.tenants[0]?.id || null] : newTenants,
                            };
                          }
                          return selectedDefinition;
                        })
                      );
                    }}
                    floating
                  />
                );
              }
            }

            return body;
          })}
          disablePagination
          noHighlight
          loading={!definitions}
          allowLocalSorting
        />
      </Modal.Content>
      <Modal.Footer>
        <Button kind="secondary" className="cancel" onClick={onClose}>
          {t('common.cancel')}
        </Button>
        <Button className="confirm" onClick={createCollection}>
          {confirmText}
        </Button>
      </Modal.Footer>
    </Modal>
  );
}

function format({key, type, tenants}: DefinitionWithTenants): Source {
  return {
    definitionKey: key,
    definitionType: type,
    tenants: tenants.map(({id}) => id),
  };
}

function definitionHasSelectedTenant(def: DefinitionWithTenants, selectedTenant?: string | null) {
  return def.tenants.some(({id}) =>
    typeof selectedTenant !== 'undefined' ? selectedTenant === id : true
  );
}
