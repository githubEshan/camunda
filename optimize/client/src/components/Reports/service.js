/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */

import {get, del, post} from 'request';
import { getFullURL } from '../../modules/api';

export async function shareReport(reportId) {
  const body = {
    reportId,
  };
  const response = await post(getFullURL(`api/share/report`), body);

  const json = await response.json();
  return json.id;
}

export async function getSharedReport(reportId) {
  const response = await get(getFullURL(`api/share/report/${reportId}`));

  if (response.status > 201) {
    return '';
  } else {
    const json = await response.json();
    return json.id;
  }
}

export async function revokeReportSharing(id) {
  return await del(getFullURL(`api/share/report/${id}`));
}
