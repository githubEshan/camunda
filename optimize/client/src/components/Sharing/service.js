/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */

import {get, post} from 'request';
import { getFullURL } from '../../modules/api';

export async function evaluateEntity(id, type, query = {}) {
  const request = type === 'dashboard' ? get : post;
  const response = await request(getFullURL(`api/share/${type}/${id}/evaluate`), {}, {query});

  return await response.json();
}

export function createLoadReportCallback(dashboardShareId) {
  return async (reportId, filter, query) => {
    const response = await post(
      getFullURL(`api/share/dashboard/${dashboardShareId}/report/${reportId}/evaluate`),
      {filter},
      {query}
    );

    return await response.json();
  };
}
