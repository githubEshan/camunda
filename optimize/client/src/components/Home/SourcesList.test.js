/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Camunda License 1.0. You may not use this file
 * except in compliance with the Camunda License 1.0.
 */

import React from 'react';
import {shallow} from 'enzyme';

import {EntityList, Deleter, Modal} from 'components';
import {addSources} from 'services';

import {getSources, removeSource, editSource} from './service';

import SourcesListWithErrorHandling from './SourcesList';
import EditSourceModal from './modals/EditSourceModal';
import SourcesModal from './modals/SourcesModal';
import {areTenantsAvailable} from 'config';

jest.mock('notifications', () => ({addNotification: jest.fn()}));
jest.mock('config', () => ({areTenantsAvailable: jest.fn().mockReturnValue(true)}));
jest.mock('services', () => ({
  ...jest.requireActual('services'),
  addSources: jest.fn(),
}));

jest.mock('./service', () => ({
  getSources: jest.fn().mockReturnValue([
    {
      id: 'process:defKey',
      definitionType: 'process',
      definitionKey: 'defKey',
      definitionName: 'definition name',
      tenants: [
        {id: null, name: 'Not defined'},
        {id: 'tenant1', name: 'Sales'},
        {id: '__unauthorizedTenantId__', name: 'Unauthorized Tenant'},
      ],
    },
    {
      id: 'process:defKey1',
      definitionType: 'process',
      definitionKey: 'defKey1',
      definitionName: 'definition 1',
      tenants: [{id: 'tenant1', name: 'Sales'}],
    },
  ]),
  removeSource: jest.fn(),
  editSource: jest.fn(),
  isUnauthorizedTenant: jest.fn(),
}));

const SourcesList = SourcesListWithErrorHandling.WrappedComponent;

const props = {
  mightFail: jest.fn().mockImplementation((data, cb) => cb(data)),
  collection: 'collectionId',
  onChange: jest.fn(),
  readOnly: false,
};

it('should match snapshot', async () => {
  const node = shallow(<SourcesList {...props} />);
  await node.update();

  expect(node).toMatchSnapshot();
});

it('should hide add button and edit menu when in readOnly mode', async () => {
  const node = shallow(<SourcesList {...props} readOnly />);
  await node.update();

  expect(node.find(EntityList).prop('action')).toBe(false);

  expect(node.find(EntityList).prop('rows')[0].actions).toEqual([]);
});

it('should pass entity to Deleter', async () => {
  const node = shallow(<SourcesList {...props} />);
  await node.update();

  node.find(EntityList).prop('rows')[1].actions[1].action();

  expect(node.find(Deleter).prop('entity').id).toBe('process:defKey1');
});

it('should delete source', () => {
  const node = shallow(<SourcesList {...props} />);

  node.setState({deleting: {id: 'sourceId'}});

  node.find(Deleter).prop('deleteEntity')();

  expect(removeSource).toHaveBeenCalledWith('collectionId', 'sourceId');
});

it('should show an edit modal when clicking the edit button', async () => {
  const node = shallow(<SourcesList {...props} />);
  await node.update();

  node.find(EntityList).prop('rows')[0].actions[0].action();

  expect(node.find(EditSourceModal)).toExist();
});

it('should modify the tenants in source', () => {
  const node = shallow(<SourcesList {...props} />);

  node.setState({editing: {id: 'sourceId'}});
  const updatedTenants = [{id: 'newTenant', name: 'New Tenant'}];
  node.find(EditSourceModal).prop('onConfirm')(updatedTenants);

  expect(editSource).toHaveBeenCalledWith('collectionId', 'sourceId', updatedTenants, undefined);
});

it('should add sources when SourcesModal is confirmed', () => {
  const node = shallow(<SourcesList {...props} />);

  node.setState({addingSource: true});
  const sources = [
    {
      id: 'sourceId',
      definitionType: 'process',
      definitionKey: 'defKey',
      tenants: ['tenantId'],
    },
  ];
  node.find(SourcesModal).prop('onConfirm')(sources);

  expect(addSources).toHaveBeenCalledWith('collectionId', sources);
});

it('should hide edit and tenants in source items if there are not tenants available', async () => {
  areTenantsAvailable.mockReturnValueOnce(false);
  const node = shallow(<SourcesList {...props} />);
  await node.update();
  expect(node.find(EntityList).props().rows[1]).toMatchSnapshot();
});

it('should pass conflict to confirmation modal if update failed', async () => {
  const conflictedItems = [{id: 'reportId', type: 'report', name: 'Report Name'}];
  const mightFail = (_promise, _cb, err) => {
    if (err) {
      err({status: 409, conflictedItems});
    }
  };

  const node = shallow(<SourcesList {...props} mightFail={mightFail} />);

  node.setState({editing: {id: 'sourceId'}});
  const updatedTenants = [{id: 'newTenant', name: 'New Tenant'}];
  node.find(EditSourceModal).prop('onConfirm')(updatedTenants);

  await node.update();
  await flushPromises();

  expect(node.find(Modal.Content)).toIncludeText(conflictedItems[0].name);

  node.setProps({mightFail: props.mightFail});
  node.find('.confirm').simulate('click');

  expect(editSource).toHaveBeenCalledWith('collectionId', 'sourceId', updatedTenants, true);
  expect(getSources).toHaveBeenCalled();
  expect(props.onChange).toHaveBeenCalled();
});
