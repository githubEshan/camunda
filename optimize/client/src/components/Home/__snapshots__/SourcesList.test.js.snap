// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should hide edit and tenants in source items if there are not tenants available 1`] = `
Object {
  "actions": Array [
    Object {
      "action": [Function],
      "icon": <ForwardRef(TrashCan) />,
      "text": "Remove",
    },
  ],
  "className": "process",
  "entityType": "process",
  "icon": <ForwardRef(TableSplit) />,
  "id": "process:defKey1",
  "meta": Array [
    false,
  ],
  "name": "definition 1",
  "type": "Process",
}
`;

exports[`should match snapshot 1`] = `
<div
  className="SourcesList"
>
  <EntityList
    action={
      <Button
        kind="primary"
        onClick={[Function]}
      >
        Add
      </Button>
    }
    bulkActions={
      Array [
        <BulkDeleter
          checkConflicts={[Function]}
          conflictMessage="Removing data sources deletes all associated reports and their alerts."
          deleteEntities={[Function]}
          type="remove"
        />,
      ]
    }
    emptyStateComponent={
      <EmptyState
        actions={
          <Button
            kind="primary"
            onClick={[Function]}
            size="md"
          >
            Add
          </Button>
        }
        description="You can add one or more sources to the collection"
        icon={<ForwardRef(Db2Database) />}
        title="There are no sources added yet"
      />
    }
    headers={
      Array [
        "Definition name",
        "Tenants",
      ]
    }
    isLoading={false}
    onChange={[Function]}
    rows={
      Array [
        Object {
          "actions": Array [
            Object {
              "action": [Function],
              "icon": <ForwardRef(Edit) />,
              "text": "Edit",
            },
          ],
          "className": "process",
          "entityType": "process",
          "icon": <ForwardRef(TableSplit) />,
          "id": "process:defKey",
          "meta": Array [
            "Not defined, Sales, (Unauthorized tenant)",
          ],
          "name": "definition name",
          "type": "Process",
        },
        Object {
          "actions": Array [
            Object {
              "action": [Function],
              "icon": <ForwardRef(Edit) />,
              "text": "Edit",
            },
            Object {
              "action": [Function],
              "icon": <ForwardRef(TrashCan) />,
              "text": "Remove",
            },
          ],
          "className": "process",
          "entityType": "process",
          "icon": <ForwardRef(TableSplit) />,
          "id": "process:defKey1",
          "meta": Array [
            "Sales",
          ],
          "name": "definition 1",
          "type": "Process",
        },
      ]
    }
  />
  <Deleter
    checkConflicts={[Function]}
    deleteEntity={[Function]}
    deleteText="Remove Data source"
    descriptionText="The '' data source will be removed from the collection."
    entity={null}
    getName={[Function]}
    onClose={[Function]}
    onDelete={[Function]}
    type="source"
  />
  <Modal
    className="saveModal"
    onClose={[Function]}
    open={null}
  >
    <ForwardRef(ModalHeader)
      title="Destructive changes"
    />
    <ForwardRef />
    <ForwardRef(ModalFooter)>
      <Button
        className="close"
        disabled={false}
        kind="secondary"
        onClick={[Function]}
      >
        Don't save
      </Button>
      <Button
        className="confirm"
        disabled={false}
        onClick={[Function]}
      >
        Save changes
      </Button>
    </ForwardRef(ModalFooter)>
  </Modal>
</div>
`;
